### PR TITLE
ARROW-9099: [C++][Gandiva] Implement trim function for string

### PR DESCRIPTION
**Pull request by sgnkc**

### CODE REVIEW

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -284,6 +284,42 @@ const char* reverse_utf8(gdv_int64 context, const char* data, gdv_int32 data_len
...
+  while (end >= start && data[end] == ' ') {
+    --end;
+  }
+
+  // string with all spaces
+  if (start > end) {
+    *out_len = 0;
+    return "";
+  }
+
```
Comment (liyafan82) [inline cpp/src/gandiva/precompiled/string_ops.cc:310]: Do we need a quick path that return the original string if there are no leading/trailing spaces?

File: cpp/src/gandiva/precompiled/string_ops_test.cc [hunk 2]
```diff
@@ -426,6 +426,33 @@ TEST(TestStringOps, TestReverse) {
...
+  out_str = trim_utf8(ctx_ptr, " Test  çåå†bD   ", 21, &out_len);
+  EXPECT_EQ(std::string(out_str, out_len), "Test  çåå†bD");
+  EXPECT_FALSE(ctx.has_error());
+
+  out_str = trim_utf8(ctx_ptr, "", 0, &out_len);
+  EXPECT_EQ(std::string(out_str, out_len), "");
+  EXPECT_FALSE(ctx.has_error());
+
+  out_str = trim_utf8(ctx_ptr, "      ", 6, &out_len);
+  EXPECT_EQ(std::string(out_str, out_len), "sadfsdgfh");
```
Comment (kiszk) [inline cpp/src/gandiva/precompiled/string_ops_test.cc:452]: Is this result correct?

File: cpp/src/gandiva/precompiled/string_ops.cc (hunk 1 above)
Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/string_ops.cc:317]: Makes sense. Added the logic.

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/string_ops.cc:317]: Makes sense. Added logic here.

File: cpp/src/gandiva/precompiled/string_ops_test.cc (hunk 2 above)
Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/string_ops_test.cc]: Thanks for pointing this out. Must have been a typo.

**Comment by pprudhvi** [comment]
LGTM

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 3]
```diff
@@ -284,6 +284,49 @@ const char* reverse_utf8(gdv_int64 context, const char* data, gdv_int32 data_len
...
+  gdv_int32 length = end - start + 1;
+  char* ret = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, length));
```
Comment (pravindra) [inline cpp/src/gandiva/precompiled/string_ops.cc:319]: why malloc and copy ? you could do 

*out_len = length;
return data + start;

File: cpp/src/gandiva/precompiled/string_ops_test.cc (hunk 2 above)
Comment (pravindra) [reply cpp/src/gandiva/precompiled/string_ops_test.cc]: this test is fixed now.

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 4]
```diff
@@ -284,6 +284,49 @@ const char* reverse_utf8(gdv_int64 context, const char* data, gdv_int32 data_len
...
+    return "";
+  }
+
+  gdv_int32 start = 0, end = data_len - 1;
+  // start and end denote the first and last positions of non-space
+  // characters in the input string respectively
+  while (start <= end && data[start] == ' ') {
+    ++start;
+  }
+  while (end >= start && data[end] == ' ') {
```
Comment (pravindra) [inline cpp/src/gandiva/precompiled/string_ops.cc:302]: for utf8 multi-byte characters, is it possible that the last byte looks like a space but it's actually part of a multi-byte character ?

Comment (kiszk) [reply cpp/src/gandiva/precompiled/string_ops.cc:302]: np
0x20 is always single-byte space. c.f. https://en.wikipedia.org/wiki/UTF-8

Comment (sagnikc-dremio) [reply cpp/src/gandiva/precompiled/string_ops.cc:302]: Verified that for utf8, 0x20 is always a single-byte space and is not part of any multi-byte characters.

**Review by pravindra — APPROVED** [review]

**Pull request closed without merging**